Bloom.md

```diff
@@ -0,0 +1,248 @@
+| 10  | 2047                                  |
+| 15  | 65535                                 |
+| 20  | 2097151                               |
+| 25  | 67108863                              |
+| 30  | 2147483647                            |
+| 35  | 68719476735                           |
+| 40  | 2199023255551                         |
+| 45  | 70368744177663                        |
+| 50  | 2251799813685247                      |
+| 55  | 72057594037927934                     |
+| 60  | 2305843008139952126                   |
+| 65  | 73786976294838205662                  |
+| 70  | 2361183241434822608686                |
+| 75  | 75557862452714323477986               |
+| 100 | 2535301200456458804968143894986       |
+
+Practically, we can expect scaling to stop well before the 50 range. With the default capacity of 100,000 and the default expansion rate of 2, the total items inserted (across all filters) would exceed the 64-bit unsigned integer limit before reaching 50 filters.
+
+
+### Choice of Bloom Filter Library
+
+We evaluated the following libraries:
+* https://crates.io/crates/bloomfilter
+* https://crates.io/crates/growable-bloom-filter
+* https://crates.io/crates/fastbloom
+
+Here are some important factors when evaluating libraries:
+* Is it actively maintained?
+* Has there been any breaking changes in serialization and deserialization of bloom filters over version updates?
+* Performance
+* Popularity
+
+https://crates.io/crates/bloomfilter was chosen as it provides all the required APIs for controlling a bloom filter. This library has also been stable through the releases.
+
+Certain libraries (e.g. fastbloom) are no longer compatible with older versions (after updates) resuling in serialization / deserialization incompatibility.
+
+growable-bloom-filter is a bloom filter library that implements a scalable bloom filter. However, it does not allow control over when the scaling occurs and does not expose APIs to check the number of filters that it has scaled out to. This crate also has had a breaking change across versions and bloom filters of the older version are not loadable on the newer versions.
+
+
+### Large BloomFilter objects
+
+Create and Delete operations on large bloom filters take longer durations and will block the main thread for this duration. 
+Because of this, the following operations can be handled differently:
+* defrag callback: If the memory used is greater than X MB (configurable with `bf.bloom_large_item_threshold`), we can skip defrag operations on this bloom object.
+* free_effort callback: This callback decides the free effort for the bloom object. If it is greater than X MB (configurable with `bf.bloom_large_item_threshold`), we can return 0 to use async free logic.
+* create operations (BF.ADD/MADD/INSERT): We can compute the estimated size of the bloom filter that will be created as a result of the operation. We can consider rejecting requests in case of unavailable memory.
```

Comment:
How long can BF.INSERT really take?

Reply:
Yes, Scalable Bloom filters will grow in used memory after creation - but only as a result of a BF.ADD, BF.MADD, or BF.INSERT operation. We can consider rejecting these requests if we identify that it will result in a scale out operation OR beyond a certain number of filters